Checkpoint

koyagabriel opened the pull request:
#### What does this PR do?
Have creation and allocation functions in the Space Allocator Application
#### Description of Task to be completed?
Have the following functions working.
- create room or rooms
- add person to a room
- print members of a room
- prints unallocations 
- prints allocations
#### How should this be manually tested?
1. To run the app itself:
- Install `docopt module` using `pip install docopt==0.6.2` for python 2 and `pip3 install docopt==0.6.2` for python 3.
- After cloning the repo,  run `python main.py`.
- To create a room run `create_room type_of_room(office/livingspace) room_name`
  - e.g  `create_room office Orange`
- To create multiple rooms at once run `create_room type_of_room(office/livingspace) room_name1 room_name2 room_name3`
  - e.g  `create_room livingspace Orange Red Green White`
- To add a person to a room run `add_person surname firstname person_type(staff/fellow) wants_accomodation(y/n)
  - e.g `add_person John Doe staff y`
- To print members of a room run `print_room room_name`
  - e.g `print_room Orange`
- To print allocations to command line run `print_allocations`
  - e.g `print_allocations`
- To print allocations to a text file run `print_allocations --o=filename`
  - e.g `print_allocations --o=data`
- To print unallocated to command line run `print_allocations`
  - e.g `print_unallocated`
- To print unallocated to a text file run `print_allocations --o=filename`
  - e.g `print_unallocated --o=data`

2. To test the app:
- Install `nosetest module` using using `pip install nose` for python 2 and `pip3 install nose` for python 3.
- After cloning the repo, cd into the `app/tests` directory and run `nosetests dojo_tests.py`.

Commit:
doctot library successfull installed

Commit:
Local changes

Commit:
Create room test file implemented

Commit:
Initial commit

Commit:
[Test] Create Room Function
Implementation of test logic for the
create room functionality

Commit:
Interactivity
Implemented an interactive
command line interface

Commit:
Create Room
Implementation of the create
room functionality

Commit:
Add Person
Implemtation of the add person
functionality for the dojo class

Commit:
Utilities Class
Creation of utilities class.
Implementation of check format
functionality.
Implementation of the unique id
generation functionality

Commit:
Implementation of dojo print room function

Commit:
Implementation of Dojo print allocation function

Commit:
Implements the print unallocated functionality

Commit:
Implements the add rom logic for the room class

Commit:
Implements the allocate room logic for the room class

Commit:
Implements the random selection of room logic

Commit:
Implements the available room generation logic

Commit:
Implements the logic for generating room members

Commit:
Implements the room exist logic

Commit:
Add Room Member
Implements the logic for adding
a member to a room

Commit:
Get Allocations
Implements the logic for getting
rooms and the persons allocated
to them

Commit:
Office class
Implements the add to office list logic
Implements the logic to check the existence of an office

Commit:
Livingspace class
Implements the add to livingspace list logic
Implements the logic to check the existence of an livingspace

Commit:
Person class Implementation
Implements the assign office logic.
Implements the get unallocated logic.
Implements the logic that checks the existence of a person

Commit:
Implements Fellow Class
Implements the logic for adding to fellow list.
Implements the logic for adding unallocated fellow to list.
Implements the logic to get unallocated fellows.

Commit:
Implements the staff class logic

Commit:
Implements functionalities for the file class

Commit:
Updates Readme.md

hassan02 commented on app/dojo.py:
```diff
@@ -0,0 +1,116 @@
+		pass
+
+
+	def create_room(room_type, room_names):
```
The create_room function is kinda long. Try refactoring the function.

hassan02 commented on app/dojo.py:
```diff
@@ -0,0 +1,116 @@
+
+	def print_room(room_name):
+		try:
+			print()
```
What is the purpose of this line

hassan02 commented on app/dojo.py:
```diff
@@ -0,0 +1,116 @@
+				print(e)
+
+	def reallocate_person(identifier, new_room_name):
+		Room.reallocate_room(identifier, new_room_name)
```
Are you reallocating room or person? Consider changing `Room.reallocate_room(identifier, new_room_name)` to `Room.reallocate_person(identifier, new_room_name)`

hassan02 commented on app/room.py:
```diff
@@ -0,0 +1,154 @@
+	def add_room(room):
+		Room.room_list[room.name] = room
+
+	def allocate_room(person, office_list, livingspace_list = None ):
```
Function `allocate_room()` too long. Consider refactoring.

hassan02 commented on app/tests/dojo_tests.py:
```diff
@@ -0,0 +1,100 @@
+		later_room_count = Dojo.get_total_rooms()
+		self.assertEqual(initial_room_count, later_room_count)
+
+	# def test_remove_room(self):
```
Remove commented code

hassan02 commented on app/tests/person_test.py:
```diff
@@ -0,0 +1,5 @@
+import unittest
```
Any reason why this file exist.

Commit:
Refactors add person function

Commit:
Refactors print unallocation logic

mfalade commented on app/dojo.py:
```diff
@@ -0,0 +1,83 @@
+class Dojo(object):
+	"""docstring for Dojo"""
+
+	def __init__(self):
```
This __init__ method is redundant, feel free to remove it

koyagabriel replied:
done

mfalade commented on app/dojo.py:
```diff
@@ -0,0 +1,83 @@
+			else:
+				print("{}\n{}".format(Office.allocate_office(person),
+									  LivingSpace.allocate_livingspace(person)))
+		except Exception as e:
```
Broad except is usually frowned upon.. You need to specify the exact exception you are trying to catch.

e.g 

```python
try:
    foo = 5/0
except ZeroDivisionError:
    print('Sorry, but you cant divide by zero')
```

koyagabriel replied:
done

mfalade commented on app/dojo.py:
```diff
@@ -0,0 +1,83 @@
+
+	def allocate_room(person, wants_accomodation=None):
+		try:
+			if((wants_accomodation is None) or (wants_accomodation.lower() != "y")):
```
This could be further simplified as 
```python
    if not wants_accomodation or wants_accomodation in ['y', 'Y']:
        do_something()
```

koyagabriel replied:
done

mfalade commented on app/dojo.py:
```diff
@@ -0,0 +1,83 @@
+			print(allocations)
+		else:
+			try:
+				file = File.create_file(filename)
```
[IMPORTANT - 🔴 ]
`file` is a built in python keyword and should not be used as a variable name

koyagabriel replied:
done

Commit:
Optimizes Logic

Commit:
Refactors logic

Commit:
Refactors logic

Commit:
Implements Reallocation logic

Commit:
Implements more test

Commit:
Refactors logic

Commit:
Implemented more test functionalities

Commit:
Fixes test bugs

Commit:
Refactored code

Commit:
Fixes reallocate function bugs

njirap commented on README.md:
```diff
@@ -1 +1,26 @@
\ No newline at end of file
+# Office-Space-Allocator
+
+#### How should this be manually tested?
```
A good way to start a README is by telling us what this project is and what it does. Then we can go on to how to set it up `locally`..
the word `manually` here may not sound professional maybe use `locally`

koyagabriel replied:
Thanks for the feedback.. I have updated it.

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+            else:
+                print("Invalid type of person")
+
+        except WrongFormatException as e:
```
under which condition(s) would this Exception be raised?

koyagabriel replied:
the exception is catching exception that might emanate from line 35 or 44 when instantiating a staff or fellow. It checks if whether the surname or first name is not made up of special characters only. The exception is raised in the Person class

class Person():
 +	"""docstring for Person"""
 +
 +	id_map = {}
 +
 +	def __init__(self, lname, fname):
 +		try:
 +			Utilities.check_format_validity([lname, fname])
 +		except WrongFormatException:
 +			raise WrongFormatException(
 +				"Firstname or Lastname is not a valid name format")
 +
 +		self.id = Utilities.generate_person_id(Person.get_id_list())
 +		self.surname = lname.capitalize()
 +		self.firstname = fname.capitalize()
 +		self.office = None

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+
+    @classmethod
+    def allocate_room(cls, person, wants_accomodation=None):
+        if(not(wants_accomodation in ['y', 'Y'])):
```
what about `if not wants_accomodation.lower() == 'y'`?

koyagabriel replied:
`if not wants_accomodation.lower() == 'y'` will throw an error if no argument is passed for the `wants_accomodation` parameter, because then, `wants_accomodation` will be `None` and the `lower()` method can't be invoked on it.

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+            try:
+                new_file = File.create_file(filename)
+                File.write(new_file, allocations)
+            except FileExistsError as e:
```
So if the file already exists we don't want to consider overwriting it?

koyagabriel replied:
Will implement the overwrite functionality

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+
+    @classmethod
+    def reallocate_person(cls, identifier, room_name):
+        person_id = None
```
why is the purpose of assigning this to None?

koyagabriel replied:
fixed the redundant statement

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+            if Person.exist(person_id):
+                person = Person.id_map[person_id]
+                if Room.exists(room_name):
+                    room = Room.get_room_list()[room_name]
```
what does `get_room_list()` in `Room` return? It's name suggests the it should return a `list`.. `room_name` implies that it's a variable of type string.

So what is going on in this line? It appears you are getting an index `room_name` of list `get_room_list()`.

List indices must be integers or slices, not string like `room_name`

koyagabriel replied:
`Room.get_room_list()` returns a dictionary, `[room_name]` is the key of the value i want to extract but notwithstanding i have refactored it. I implemented list comprehension in its replacement, so as to eliminate using a dictionary containing self referencing objects as key-value pair and also to follow standard variable naming convention because `get_room_list()` should return a list and not a dictionary.

Thanks for the feedback and insight.

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+                                           + "can't be assigned to a staff")
+                                          .format(room_name))
+                            else:
+                                if room.get_type() == "office":
```
What is the difference with Line 109 here?

njirap replied:
seen the difference..

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,164 @@
+                              + "because of incomplete information\n")
+        try:
+            file_to_be_loaded = File.open_file(filename)
+            with file_to_be_loaded as data_file:
```
please explain what is going on here..

koyagabriel replied:
i created a list to hold all the error messages that might emanate from adding people via loading a file. The possible errors are incomplete arguments. The list holds error messages for every line in the file that throws an error which i later outputted all together in a well formatted style.
I have refactored line 139 to line 142 in order to eliminate multiple calls to the append method.

njirap commented on app/fellow.py:
```diff
@@ -0,0 +1,63 @@
+from app.person import Person
+
+sys.path.append(path.dirname(path.dirname(
+    path.dirname(path.abspath(__file__)))))
```
please explain what is going on here..

koyagabriel replied:
Fixed the redundant statement

njirap commented on app/fellow.py:
```diff
@@ -0,0 +1,63 @@
+    """docstring for Fellow"""
+
+    fellow_list = []
+    unallocated_fellows = {"office": [], "livingspace": []}
```
is this supposed to be `unallocated` or `allocated`?

koyagabriel replied:
it holds the unallocated fellows. Those yet to have an office will be appended to the office section of the dictionary, while those yet to get a livingspace will be appended to the livingspace section of the dictionary and those yet to get both office and livingspace will be appended to both sections of the dictionary.

njirap commented on app/fellow.py:
```diff
@@ -0,0 +1,63 @@
+        self.category = new_category.lower()
+
+    def get_category(self):
+        return self.category
```
checkout Python getters and setters..

http://www.python-course.eu/python3_properties.php

koyagabriel replied:
okay. Thanks... I have implemented them.

njirap commented on app/file.py:
```diff
@@ -0,0 +1,45 @@
+from os import path
```
`file` is a type in Python 2...Inasmuch as this is Py3 it is still not professional to call this file `file.py` because in the case of Py2 having `from app import file` would overwrite the builtin `file` type

koyagabriel replied:
bug fixed. package and class name changed to **Customfile**

njirap commented on main.py:
```diff
@@ -0,0 +1,101 @@
+	except SystemExit:
+		pass
+	except KeyboardInterrupt:
+		pass
```
hmmm.. please expound on what is going on here..

koyagabriel replied:
Removed the unnecessary try/except block.

njirap commented on main.py:
```diff
@@ -0,0 +1,101 @@
+	def do_reallocate_person(self, arg):
+		"""Usage: reallocate_person <person_identifier> <new_room_name> """
+		Dojo.reallocate_person(
+			arg['<person_identifier>'], arg['<new_room_name>'].capitalize())
```
why are we capitalising here?

koyagabriel replied:
Removed redundant method call. Thanks.

njirap commented on app/utilities.py:
```diff
@@ -0,0 +1,20 @@
+
+
+class Utilities():
+    """docstring for Utilities"""
```
is this a valid docstring?

koyagabriel replied:
Rectified the error with a valid docstring

njirap commented on app/tests/dojo_tests.py:
```diff
@@ -0,0 +1,215 @@
+        new_office = Dojo.create_room("office", ["red"])[0]
+        Dojo.add_person("koya", "gabriel", "staff")
+        person = Staff.get_staff_list()[0]
+        self.assertEqual(len(new_office.get_room_members()), 1)
```
before we can assert this maybe after L204 we should assert that the room is empty..

koyagabriel replied:
Done. Thanks

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+        if room_name.capitalize() in list(cls.room_list.keys()):
+            return True
+
+        return False
```
This statement basically is:
```
if True:
    return True
return False
```

There is no need for the `if` statement if we can have this as

`return room_name.capitalize() in list(cls.room_list.keys())`

the statement in the `if` already returns a Boolean so there is no need to instruct it to return a Boolean equivalent to the Boolean it has returned.

koyagabriel replied:
Implemented the feedback. Thanks for the insight.

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+        return cls.total_number_of_rooms
+
+    def get_capacity_used(self):
+        return self.capacity_used
```
look into python getters and setters..

koyagabriel replied:
Done

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+            if not room.is_full:
+                available_rooms.append(room)
+
+        return available_rooms
```
do you know about comprehensions in python? In this case `list comprehension`

this function can have 1 line of code as such
`return [room for room in room_list if not room.is_full]`

koyagabriel replied:
Implemented the comprehension. Thanks for the feedback and insight.

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+class Room():
+
+    total_number_of_rooms = 0
+    room_list = {}
```
the name `room_list` implies that it is a list...this variable name is confusing..

koyagabriel replied:
i have refactored it into a  list in order to uphold variable naming convention. Thanks for the feedback

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+        if not isinstance(room, Room):
+            print("Only a room, office, or livingspace instance can be added")
+            return False
+        cls.room_list[room.name] = room
```
hmmm...this does not look professional..Having a map with the key being one of the values of an instance while the value being the entire instance itself..

This does not look any different from mapping something to itself..

Maybe `room_list` can just be a list of room instances instead..

koyagabriel replied:
I have implemented a list and removed the self mapping dictionary.

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+            return False
+        self.room_members.append(person)
+        if len(self.room_members) == self.maximum_capacity:
+            self.is_full = True
```
my opinion is that we don't need a dedicated boolean variable to tell us if the room is full because we can always get this boolean from comparing `len(self.room_members) == self.maximum_capacity`

koyagabriel replied:
Yes you are right and it will reduce unnecessary memory consumption reducing space complexity. I have removed the boolean variable and implemented the boolean condition expression in its place. Thanks for the insight.

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+
+    def __init__(self, room_name):
+        self.name = room_name.capitalize()
+        self.is_full = False
```
refer to comment on L37..

koyagabriel replied:
Made necessary corrections. Thanks for the feedback.

njirap commented on app/room.py:
```diff
@@ -0,0 +1,116 @@
+    room_list = {}
+
+    def __init__(self, room_name):
+        self.name = room_name.capitalize()
```
what is the value of capitalising this?

koyagabriel replied:
I decided to capitalize it for better display format when outputting a message to the user. I decided to capitalize it when invoking the `init()` so as to eliminate repetitive call to `capitalize()` whenever i want to output a message to the user.

njirap commented on app/person.py:
```diff
@@ -0,0 +1,92 @@
+	def exist(cls, person_identifier):
+		if not person_identifier in cls.get_id_list():
+			return False
+		return True
```
this can be reduced to 1 line `return person_identifier in cls.get_id_list()`

koyagabriel replied:
Done.. Thanks

njirap commented on app/person.py:
```diff
@@ -0,0 +1,92 @@
+
+class Person():
+	"""docstring for Person"""
+
```
have you used tabs or spaces in this file.. whatever you have used be sure to be consistent to use the same  thing throughout all the files in this project.. the indentations in this file don't match with some other files..

koyagabriel replied:
i have been using tabs. I have fixed the indentation mismatch.

njirap commented on app/person.py:
```diff
@@ -0,0 +1,92 @@
+		self.office = office
+
+	def get_assigned_office(self):
+		return self.office
```
look into python getters and setters..

koyagabriel replied:
okay. Thanks.. Implemented them

Commit:
Code Refactoring

Commit:
applies getters and setters to code

Commit:
Added extra assert statement to test

Commit:
Refactors boolean condition expression

Commit:
Fixes bug in dojo class

Commit:
Refactors code

Commit:
Remove redundant comments

Commit:
Refactors code
Makes us of list comprehension rather than mapping an object to itself
using a dict

Commit:
Removes irrelevant comments

Commit:
Adds more test cases

Commit:
Refactors code, replacing all self referencing dictionaries with list comprehension

Commit:
Refactors code
self mapping of dictionaries, where replaced with list comprehension

Commit:
Refactors code

Commit:
Refactors code

Commit:
Refactors code

Commit:
Adds more unittest

Commit:
Adds more unittest

Commit:
Added more unittest

Commit:
Adds more unittest

Commit:
Adds more unittest

Commit:
Adds more unittest

Commit:
Removes redundant file

Commit:
Adds more unittest

Commit:
Refactors unittest

Commit:
Adds more unittest

Commit:
Adds more unittest

Commit:
Adds rename room functionalities

Commit:
Refactors code

Commit:
Adds doc strings to the room class

Commit:
Includes doc string comments

Commit:
Adds override and append to file functionality

Commit:
Refactors Code

Commit:
Updates Readme

Commit:
Updates readme

Commit:
Updates readme

Commit:
creates travis.yml

Commit:
integrate .travis.yml

Commit:
integrate coverall

Commit:
integrate coverall

njirap commented on app/dojo.py:
```diff
@@ -0,0 +1,344 @@
+	# 			print("Person with the id {} doesn't exist".format(person_id))
+	#
+	# 	except ValueError:
+	# 		print("Wrong id format. id must be a number")
```
don't push commented code..

Commit:
Code Refactoring

njirap commented on app/customfile.py:
```diff
@@ -5,13 +5,11 @@
 class Customfile():
-	"""This class holds the properties and methods of a customfile
-	model """
+	"""This class holds the properties and methods of a customfile model """
```
```
"""
Docs goes here
"""
```

njirap commented on app/customfile.py:
```diff
@@ -5,13 +5,11 @@
 	def create_file(cls, filename, append_flag=False, override_flag=False):
-		""" this method creates a new file using the filename
-		passed as argument """
+	""" this method creates a new file using the filename passed as argument """
```
`# this method creates a new file using the filename passed as argument`
